app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,156 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.home.blocklist
+
+import androidx.annotation.VisibleForTesting
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.support.ktx.kotlin.sha1
+import org.mozilla.fenix.home.HomeFragmentAction
+import org.mozilla.fenix.home.HomeFragmentState
+import org.mozilla.fenix.home.recentbookmarks.RecentBookmark
+import org.mozilla.fenix.home.recenttabs.RecentTab
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem
+import org.mozilla.fenix.utils.Settings
+
+/**
+ * This [Middleware] reacts to item removals from the home screen, adding them to a blocklist.
+ * Additionally, it reacts to state changes in order to filter them by the blocklist.
+ *
+ * @param settings Blocklist is stored here as a string set
+ */
+class BlocklistMiddleware(
+    private val settings: Settings
+) : Middleware<HomeFragmentState, HomeFragmentAction> {
+
+    private var blocklist: Set<String> = settings.homescreenBlocklist
+
+    /**
+     * Will filter "Change" actions using the blocklist and use "Remove" actions to update
+     * the blocklist.
+     */
+    override fun invoke(
+        context: MiddlewareContext<HomeFragmentState, HomeFragmentAction>,
+        next: (HomeFragmentAction) -> Unit,
+        action: HomeFragmentAction
+    ) {
+        val state = context.state
+        next(
```

Comment:
nit: I think `next( big block to transform action )` is hard to read. I would add a method or move out:

```kotlin
val filteredAction = applyBlockList(action)
next(filteredAction)
```

or

```kotlin
next(applyBlockList(action))
```

Reply:
This has been cleaned up